enhancements/network/ovn-kubernetes-multiple-external-gateway.md

```diff
@@ -0,0 +1,180 @@
+- Modify the existing logic to make it CRD aware, so that it can watch for changes in an annotation, as it works today, or fallback to the CR, in case the annotation does not exist.
+- Extend test coverage to include the new CRD use cases, matching the same use cases covered in the annotation tests.
+- CRD configuration will be the only officially supported and documented mechanism in OpenShift >= 4.13. Annotations will still be available in OpenShift <= 4.13.
+- Extend the existing unit tests for the annotation based logic to accommodate for the scenario where CR and annotations exist in 4.13.
+- Add a new field per hop that determines if the Source NAT is to be disabled for that gateway.
+
+### Non-Goals
+
+- Removal of the annotation based logic to configure multiple external gateways. This should be carrier out at least no sooner than OpenShift 4.14 to guarantee a stable deprecation path to existing deployments that leverage on the annotations to configure the external gateways.
+
+## Proposal
+
+Create a new cluster scoped CRD that handles internal and external routing scenarios. For the purpose of this enhancement, only the external use case will be covered.
+
+The implementation will also include a new logic that will watch for instances of this CRD and apply the changes to the OVN network configuration as it is currently done by the annotation logic. 
+
+To accommodate the fact that currently there is an annotation based logic that fulfills the external gateway functionality and that they can be both enabled in OpenShift 4.13, the logic will use CR instances as the primary source, even if an annotation exists that targets the same namespace.
+It will only revert to use the annotation configuration when there are no CR instances that target a given namespace.
+
+The annotation logic will be removed in OpenShift 4.14.
+
+### Workflow Description
+The workflow follows the behavior of a cluster scoped CR.
+
+
+### API Extensions
+
+A new cluster scoped CRD `AdminPolicyBasedExternalRoute` is created to represent the multiple external gateways configuration. The `spec` structure exposes the following fields:
+
+* `policies`: slice of policies:
+      * `from`: determines where the routing polices should be applied to. Only `namespaceSelector` is supported for external traffic. This field is mandatory.
+      * `nexthops`: defines the destination where the packets should be forwarded to. There are two types of references: static and dynamic. Static IPs usually refer to interfaces that hold an IP which doesn't change overtime. Dynamic IPs link to pod IPs and can potentially change over time as they are linked to the pod's lifecycle.
+        *`static`: is a slice of static IPs, each item containing the following fields:
+            * `ip`: IPv4 or IPv6 of the next destination hop. This field is mandatory.
+            * `skipHostSNAT`: When true, it disables applying the Source NAT to the host IP.
```

Comment:
why do we need this `skipHostSNAT` ? Don't we have that info on the `from` field? all pods in that namespace won't have the SNAT configured right?

Does this feature actually support having a SNAT towards hostIP? -> I thought this is something that can be deciphered from the `disableSNATMultipleGWs` option. Wondering why we need this flag here on a per externalgw basis? cc @trozet ?

Reply:
Another thought that just came in was both `skipHostSNAT` and `bfdEnabled`, these seem to be set for the whole set of static exgw pods or dynamic exw pods.. should this be allowed on a per exgw basis? -

because for the namespace annotations its true we behave this way but for pod annotations, we actually provide a way to do this per exgw pod.

Reply:
disableSNATMultipleGWs was a poor naming choice by me. Really all it should mean is "perPodSNAT". Then with this granularity a user can decide whether or not the the pod should be SNAT'ed on a per route basis. This is an enhancement compared to the previous behavior.